src/coreclr/src/vm/callcounting.cpp

```diff
@@ -0,0 +1,1735 @@
+{
+    WRAPPER_NO_CONTRACT;
+    return true;
+}
+
+#endif // !X64
+
+////////////////////////////////////////////////////////////////
+// X64 || X86
+
+#if defined(_TARGET_AMD64_) || defined(_TARGET_X86_)
+
+#ifndef DACCESS_COMPILE
+template<class T> INT_PTR CallCountingManager::CallCountingStub::GetRelativeOffset(const T *relRef, PCODE target)
+{
+    WRAPPER_NO_CONTRACT;
+    static_assert_no_msg(sizeof(T) != 0);
+    static_assert_no_msg(sizeof(T) <= sizeof(void *));
+    static_assert_no_msg((sizeof(T) & (sizeof(T) - 1)) == 0); // is a power of 2
+    _ASSERTE(relRef != nullptr);
+
+    TADDR targetAddress = PCODEToPINSTR(target);
+    _ASSERTE(targetAddress != NULL);
+    return (INT_PTR)targetAddress - (INT_PTR)(relRef + 1);
+}
+#endif
+
+template<class T> PCODE CallCountingManager::CallCountingStub::GetTarget(const T *relRef)
+{
+    WRAPPER_NO_CONTRACT;
+    static_assert_no_msg(sizeof(T) == 1 || sizeof(T) == 2 || sizeof(T) == 4 || sizeof(T) == 8);
+    _ASSERTE(relRef != nullptr);
+
+    return PINSTRToPCODE((INT_PTR)(relRef + 1) + *relRef);
+}
+
+#ifndef DACCESS_COMPILE
+
+const CallCountingManager::CallCountingStubShort *CallCountingManager::CallCountingStubShort::From(TADDR stubIdentifyingToken)
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(Is(stubIdentifyingToken));
+    _ASSERTE(stubIdentifyingToken % Alignment == offsetof(CallCountingStubShort, m_alignmentPadding[0]) % Alignment);
+
+    const CallCountingStubShort *stub =
+        (const CallCountingStubShort *)(stubIdentifyingToken - offsetof(CallCountingStubShort, m_alignmentPadding[0]));
+    _ASSERTE(IS_ALIGNED(stub, Alignment));
+    return stub;
+}
+
+INT32 CallCountingManager::CallCountingStubShort::GetRelative32BitOffset(const INT32 *rel32Ref, PCODE target)
+{
+    WRAPPER_NO_CONTRACT;
+
+    INT_PTR relativeOffset = GetRelativeOffset(rel32Ref, target);
+    _ASSERTE((INT32)relativeOffset == relativeOffset);
+    return (INT32)relativeOffset;
+}
+
+#endif // !DACCESS_COMPILE
+
+PCODE CallCountingManager::CallCountingStubShort::GetTargetForMethod() const
+{
+    WRAPPER_NO_CONTRACT;
+    return GetTarget(&m_rel32TargetForMethod);
+}
+
+#endif // X64 || X86
+
+////////////////////////////////////////////////////////////////
+// Arm64 || Arm32
+
+#if defined(_TARGET_ARM64_) || defined(_TARGET_ARM_)
+
+#ifndef DACCESS_COMPILE
+const CallCountingManager::CallCountingStubShort *CallCountingManager::CallCountingStubShort::From(TADDR stubIdentifyingToken)
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(Is(stubIdentifyingToken));
+
+    const CallCountingStubShort *stub = (const CallCountingStubShort *)stubIdentifyingToken;
+    _ASSERTE(IS_ALIGNED(stub, Alignment));
+    return stub;
+}
+#endif
+
+PCODE CallCountingManager::CallCountingStubShort::GetTargetForMethod() const
+{
+    WRAPPER_NO_CONTRACT;
+    return m_targetForMethod;
+}
+
+#endif // Arm64 || Arm32
+
+////////////////////////////////////////////////////////////////
+// X64
+
+#if defined(_TARGET_AMD64_)
```

Comment:
It would be better to have these platform specific details in platform specific files (putting them into the existing platform specific files e.g. `cgencpu.h` is fine)

Reply:
`callcounting.h` has too many dependencies to include into something like `cgencpu.h`. I can put the arch-specific definitions in `cgen<arch>.cpp` files. They may not be inlined on Unixes, but it might be ok. Alternatively I can separate the definitions into arch-specific and call counting-specific `.h` files that are included into the `callcounting.cpp`. I'm leaning towards the latter, let me know if the former would be preferable instead.

Reply:
My preference would be:
- Factor the platform-specific stubs so that they have minimal dependencies on the rest of the system. ie the higher level platform neutral call counting stuff should depend on the low level stuff. Not the other way around. Once you do that, there should not be a problem with having it in cgencpu.h.
- Avoid tiny files. E.g. I would merge the existing callcounting*.asm files that just contain a single method into one of the larger asm files.